# Introduction

The goal of this project is to practice cleaning a dataset. Data cleaning is a critical step in data science, ensuring that raw data is properly prepared for deeper analysis. In this project, we'll focus on identifying and correcting errors, handling missing values, and standardizing the data to ensure its quality and consistency. The dataset used in this project was sourced from Kaggle, providing a real-world example to apply these essential data cleaning techniques.


In [1]:
import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('Dataset/Wine_Quality.csv')

In [3]:
df.head()

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6497 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   type                  6497 non-null   object 
 1   fixed acidity         6487 non-null   float64
 2   volatile acidity      6489 non-null   float64
 3   citric acid           6494 non-null   float64
 4   residual sugar        6495 non-null   float64
 5   chlorides             6495 non-null   float64
 6   free sulfur dioxide   6497 non-null   float64
 7   total sulfur dioxide  6497 non-null   float64
 8   density               6497 non-null   float64
 9   pH                    6488 non-null   float64
 10  sulphates             6493 non-null   float64
 11  alcohol               6497 non-null   float64
 12  quality               6497 non-null   int64  
dtypes: float64(11), int64(1), object(1)
memory usage: 660.0+ KB


In [14]:
memory_usage = df.memory_usage(deep=True)
total_memory_usage = memory_usage.sum()

In [15]:
total_memory_usage

971484

In [5]:
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    6487.000000       6489.000000  6494.000000     6495.000000   
mean        7.216579          0.339691     0.318722        5.444326   
std         1.296750          0.164649     0.145265        4.758125   
min         3.800000          0.080000     0.000000        0.600000   
25%         6.400000          0.230000     0.250000        1.800000   
50%         7.000000          0.290000     0.310000        3.000000   
75%         7.700000          0.400000     0.390000        8.100000   
max        15.900000          1.580000     1.660000       65.800000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  6495.000000          6497.000000           6497.000000  6497.000000   
mean      0.056042            30.525319            115.744574     0.994697   
std       0.035036            17.749400             56.521855     0.002999   
min       0.009000             1.000000              6.000000     0.987110   
25%       0.038000            17.000000             77.000000     0.992340   
50%       0.047000            29.000000            118.000000     0.994890   
75%       0.065000            41.000000            156.000000     0.996990   
max       0.611000           289.000000            440.000000     1.038980   

                pH    sulphates      alcohol      quality  
count  6488.000000  6493.000000  6497.000000  6497.000000  
mean      3.218395     0.531215    10.491801     5.818378  
std       0.160748     0.148814     1.192712     0.873255  
min       2.720000     0.220000     8.000000     3.000000  
25%       3.110000     0.430000     9.500000     5.000000  
50%       3.210000     0.510000    10.300000     6.000000  
75%       3.320000     0.600000    11.300000     6.000000  
max       4.010000     2.000000    14.900000     9.000000

In [6]:
df.shape

(6497, 13)

## 1. Dealing with Missing Data 

The first step in this project is to focus on handling missing data. Missing values can distort analysis and lead to inaccurate conclusions, so it's crucial to identify and address them early in the data cleaning process.

In [7]:
df.isna().sum()

type                     0
fixed acidity           10
volatile acidity         8
citric acid              3
residual sugar           2
chlorides                2
free sulfur dioxide      0
total sulfur dioxide     0
density                  0
pH                       9
sulphates                4
alcohol                  0
quality                  0
dtype: int64

In [8]:
max_missing_values = len(df)*0.05

In [9]:
max_missing_values

324.85

We will delete all rows where more than 5% of the values are missing, which equates to approximately 325.85 entries. This approach helps ensure that the dataset remains robust and that the analysis is based on sufficiently complete records. By removing these incomplete rows, we aim to improve the overall quality of our data for subsequent analyses.

In [10]:
columns_to_check = df.columns[(df.isna().sum() < max_missing_values) & (df.isna().sum() > 0)]


In [11]:
columns_to_check

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'pH', 'sulphates'],
      dtype='object')

In [12]:
df_filtered = df.dropna(subset=columns_to_check)

In [13]:
df_filtered.isna().sum()

type                    0
fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

Since they represent only a small portion of NaN values, we can simply delete all these rows. This approach is justified because the removal of a limited number of entries will not significantly impact the overall dataset. By doing so, we can ensure that our analysis is conducted on cleaner data, which is crucial for obtaining reliable results.

## 2. Dtypes 

In [19]:
df_filtered["type"].nbytes

51704

In [16]:
df_filtered["type"].nunique()

2

In [17]:
df_filtered["type"].unique()

array(['white', 'red'], dtype=object)

We can change the data type to 'category' to **enhance** both the **comprehensibility** and **memory efficiency** of the dataset.

In [21]:
df_filtered["type"] = df_filtered["type"].astype('category')

/var/folders/lz/zsz00k2n3pbflxxlfqttyt500000gn/T/ipykernel_2388/573768755.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered["type"] = df_filtered["type"].astype('category')


In [22]:
df_filtered["type"].dtypes

CategoricalDtype(categories=['red', 'white'], ordered=False, categories_dtype=object)

In [23]:
df_filtered["type"].nbytes

6479

We have reduced the size from 51,704 bytes to 6,479 bytes, which represents a reduction of approximately **87.45%**.

## 3. Dealing with Outliers

Outliers are data points that significantly differ from the majority of a dataset, often lying far outside the expected range. 

Outliers may need to be removed because they can skew statistical analyses, distort results, lead to misleading conclusions, and adversely affect the performance of machine learning models, ultimately compromising the reliability of insights drawn from the data.

In [26]:
outliers_dict = {}

In [30]:
for col in df_filtered.select_dtypes(include='number').columns:
    Q1 = df_filtered[col].quantile(0.25)
    Q3 = df_filtered[col].quantile(0.75)
    IQR = Q3 - Q1
    
    limite_inferieure = Q1 - 1.5 * IQR
    limite_superieure = Q3 + 1.5 * IQR
    
    outliers = df_filtered[(df_filtered[col] < limite_inferieure) | (df_filtered[col] > limite_superieure)]
    
    outliers_dict[col] = outliers

    print(f"\nFor the column '{col}', the thresholds for detecting outliers are:")
    print(f"Lower limit: {limite_inferieure}")
    print(f"Upper limit: {limite_superieure}")


For the column 'fixed acidity', the thresholds for detecting outliers are:
Lower limit: 4.450000000000001
Upper limit: 9.65

For the column 'volatile acidity', the thresholds for detecting outliers are:
Lower limit: -0.024999999999999994
Upper limit: 0.655

For the column 'citric acid', the thresholds for detecting outliers are:
Lower limit: 0.03999999999999998
Upper limit: 0.6000000000000001

For the column 'residual sugar', the thresholds for detecting outliers are:
Lower limit: -7.6499999999999995
Upper limit: 17.549999999999997

For the column 'chlorides', the thresholds for detecting outliers are:
Lower limit: -0.002500000000000009
Upper limit: 0.10550000000000001

For the column 'free sulfur dioxide', the thresholds for detecting outliers are:
Lower limit: -19.0
Upper limit: 77.0

For the column 'total sulfur dioxide', the thresholds for detecting outliers are:
Lower limit: -41.5
Upper limit: 274.5

For the column 'density', the thresholds for detecting outliers are:
Lower limit

In [35]:
total_outlier_sum = 0

for col, outliers in outliers_dict.items():
    # Add the outlier values to the total sum
    total_outlier_sum += outliers[col].sum()

# Display the total sum of outlier values
print(f"\nThe total sum of all outlier values is: {total_outlier_sum}")



The total sum of all outlier values is: 18034.72058


For this step, we can conduct research to determine whether these rates are considered outliers in the field of wine.

- **Fixed Acidity**  
  - Lower limit: 4.45  
  - Upper limit: 9.65  
  - Interpretation: Fixed acidity levels in wine typically range between 4 and 10 g/L. Thus, this range is plausible for most wines.

- **Volatile Acidity**  
  - Lower limit: -0.025  
  - Upper limit: 0.655  
  - Interpretation: A level of 0.6 g/L is considered high. Values above this may indicate defects in the wine, making this limit relevant.

- **Citric Acid**  
  - Lower limit: 0.04  
  - Upper limit: 0.60  
  - Interpretation: Levels of this acid in wine are generally low. Values up to 0.6 g/L are acceptable, making this range plausible.

- **Residual Sugar**  
  - Lower limit: -7.65  
  - Upper limit: 17.55  
  - Interpretation: Wines can have a wide range of residual sugars, especially in sweet wines. Therefore, this range is plausible.

- **Chlorides**  
  - Lower limit: -0.0025  
  - Upper limit: 0.1055  
  - Interpretation: Chloride levels in wine should be low. Values above 0.1 g/L can be concerning, so this range is significant.

- **Free Sulfur Dioxide**  
  - Lower limit: -19.0  
  - Upper limit: 77.0  
  - Interpretation: Free sulfur dioxide levels must be well-controlled to avoid defects. This range is considered acceptable.

- **Total Sulfur Dioxide**  
  - Lower limit: -41.5  
  - Upper limit: 274.5  
  - Interpretation: Total sulfur dioxide levels in wine can reach up to 300 mg/L, so this range is reasonable.

- **Density**  
  - Lower limit: 0.9853  
  - Upper limit: 1.0040  
  - Interpretation: The density of red wines typically ranges from 0.990 to 1.010 g/mL, making this range plausible.

- **pH**  
  - Lower limit: 2.795  
  - Upper limit: 3.635  
  - Interpretation: pH levels in wine generally range from 2.9 to 4.0, so this range is acceptable.

- **Sulphates**  
  - Lower limit: 0.175  
  - Upper limit: 0.855  
  - Interpretation: Sulfate levels up to 0.5 g/L are typical. Thus, this range is plausible.

- **Alcohol**  
  - Lower limit: 6.8  
  - Upper limit: 14.0  
  - Interpretation: Alcohol levels in wine typically range from 8% to 15%, so this range is reasonable.



Therefore, we establish the limits and remove all data points that fall outside of these boundaries.

In [32]:
limits = {
    'fixed acidity': (4, 10),
    'volatile acidity': (-0.025, 0.655),
    'citric acid': (0.04, 0.60),
    'residual sugar': (-7.65, 17.55),
    'chlorides': (-0.0025, 0.1055),
    'free sulfur dioxide': (-19.0, 77.0),
    'total sulfur dioxide': (-41.5, 300),
    'density': (0.9853, 1.0040),
    'pH': (2.795, 4),
    'sulphates': (0.175, 0.855),
    'alcohol': (6.8, 15.0),
    'quality': (1, 10)
}

In [40]:
df_filtered.shape


(5102, 13)

In [41]:
df_filtered_copy = df_filtered.copy()

In [44]:
# Create a mask to filter the DataFrame
mask = pd.Series([True] * len(df))

# Iterate through the limits and apply the mask
for col, (lower, upper) in limits.items():
    mask &= (df[col] >= lower) & (df[col] <= upper)

# Resetting the index of the mask and DataFrame for alignment
mask.index = df.index

# Filter the DataFrame to keep only rows within the limits
df_filtered = df[mask]



In [45]:
df_filtered.shape

(5102, 13)

In [49]:
mask = pd.Series([True] * len(df_filtered_copy))

for col, (lower, upper) in limits.items():
    mask &= (df_filtered_copy[col] >= lower) & (df_filtered_copy[col] <= upper)
mask.index = df_filtered_copy.index
df_filtered_copy = df_filtered_copy[mask]

In [51]:
df_filtered_copy.shape

(3510, 13)

In [53]:
clean_df = df_filtered_copy

In [55]:
clean_df.head()

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
4   white            7.2              0.23         0.32            8.50   
5   white            8.1              0.28         0.40            6.90   
6   white            6.2              0.32         0.16            7.00   
8   white            6.3              0.30         0.34            1.60   
10  white            8.1              0.27         0.41            1.45   

    chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
4       0.058                 47.0                 186.0   0.9956  3.19   
5       0.050                 30.0                  97.0   0.9951  3.26   
6       0.045                 30.0                 136.0   0.9949  3.18   
8       0.049                 14.0                 132.0   0.9940  3.30   
10      0.033                 11.0                  63.0   0.9908  2.99   

    sulphates  alcohol  quality  
4        0.40      9.9        6  
5        0.44     10.1        6  
6        0.47      9.6        6  
8        0.49      9.5        6  
10       0.56     12.0        5

In [56]:
memory_usage = clean_df.memory_usage(deep=True)
total_memory_usage = memory_usage.sum()

In [57]:
total_memory_usage

368764

## Conclusion
In this project, we meticulously cleaned the entire dataset by addressing missing values, managing outliers, and refining column types. The primary goal of this preparatory work was to optimize the dataset for machine learning, ensuring it operates as efficiently as possible. Additionally, we reduced the file size to minimize costs associated with data storage and processing. Through these efforts, we have laid a solid foundation for further analysis and model development, enhancing the overall performance and effectiveness of our machine learning initiatives.